### PR TITLE
ported the joint_trajectory_controller query_state service to ROS2

### PR DESCRIPTION
**Pull request by saikishor**
Hello!

There is a missing query_state from the ROS1 `joint_trajectory_controllers` : https://github.com/ros-controls/ros_controllers/blob/noetic-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L662-L708. I've ported this service to the ROS2 version of `joint_trajectory_controllers`

### CODE REVIEW

**Review by destogl — COMMENTED** [review]
Thanks for your contribution :)

I understand that this is available in ROS 1, but what I am uncertain if this is really needed, since the same data are available on the `state` topic.

Can you describe a bit more purpose of this service in your application?

**Comment by saikishor** [comment]
Hello @destogl!

> I understand that this is available in ROS 1, but what I am uncertain if this is really needed, since the same data are available on the `state` topic.
> 
> Can you describe a bit more purpose of this service in your application?


Thanks for the imminent response. The idea of this query state is that you can be able to query the future state of the trajectory. For instance, you can pass a timestamp 1 second in the future, and get the state it is going to be.

I can be useful for different usecases, for instance, after sending the trajectory, I could utilise it to check collisions in future joint references and also can use it as a prediction so if my next trajectory is around that point, I can avoid sending it etc.

I hope this answers your question. :)

**Comment by bmagyar** [comment]
As we discussed on Slack, let's wait until there's a tiny refactor & a basic test on the PR

**Comment by saikishor** [comment]
> As we discussed on Slack, let's wait until there's a tiny refactor & a basic test on the PR

@bmagyar I've reiterated through the MR and modified my initial approach and simplified it much further as the earlier instance seems to have duplicated checks. Please check and let me know what you think.

**Comment by saikishor** [comment]
@bmagyar The failing tests seems to be of admittance control, and binary build seems to have issue with the rosdep. Ofcourse, these changes break ABI🤐

**Review by bmagyar — APPROVED** [review]

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 1]
```diff
@@ -462,6 +462,53 @@ bool JointTrajectoryController::read_state_from_command_interfaces(JointTrajecto
...
 }
 
+void JointTrajectoryController::query_state_service(
+  const std::shared_ptr<control_msgs::srv::QueryTrajectoryState::Request> request,
+  std::shared_ptr<control_msgs::srv::QueryTrajectoryState::Response> response)
+{
+  const auto logger = get_node()->get_logger();
+  const auto active_goal = *rt_active_goal_.readFromRT();
+  response->name = params_.joints;
+  trajectory_msgs::msg::JointTrajectoryPoint state_requested = state_current_;
```
[the remark this answers is not in the document]
Comment (saikishor) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp:472]: Agreed!
Thank you for the feedback @bmagyar

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 2]
```diff
@@ -462,6 +462,53 @@ bool JointTrajectoryController::read_state_from_command_interfaces(JointTrajecto
...
+  // Preconditions
+  if (get_state().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE)
+  {
+    RCLCPP_ERROR(logger, "Can't sample trajectory. Controller is not running.");
```
[the remark this answers is not in the document]
Comment (saikishor) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp]: Done

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (saikishor) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp]: Updated as requested. :)

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmagyar) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp]: thx

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (bmagyar) [reply joint_trajectory_controller/src/joint_trajectory_controller.cpp]: Grand!

**Comment by bmagyar** [comment]
@mergifyio backport humble

**Pull request merged**